File: sort/2_biggest_num/answer.cpp
/*
 * cmp func paramater to reference, no need to declare temp1, temp2.
 * remember the exception the 0s.
 */

#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool cmp(string a, string b)
{
    string temp1 = a + b;
    string temp2 = b + a;

    return temp1 > temp2;
}

string solution(vector<int> numbers) {
    vector<string> temp;
    string answer = "";
    for (auto n : numbers)
        temp.push_back(to_string(n));
    sort(temp.begin(), temp.end(), cmp);
    if (temp[0] == "0")
        return "0";
    for (auto t : temp)
        answer += t;
    return answer;
}

File: sort/2_biggest_num/answer.py
# import cmp_to_key from functools
# lambda cannot make extra variable (external variable needed)
# 0 exception caution

from functools import cmp_to_key

def solution(numbers):
    answer = ''
    numbers = list(map(str, numbers))
    numbers.sort(key=cmp_to_key(lambda x,y:int(y+x) - int(x+y)))
    for n in numbers:
        answer += n
    if (numbers[0] == "0"):
        return "0"
    return answer
